File: flow/history.py
"""Private, local SQLite storage for recent transcription text."""

from dataclasses import dataclass
from datetime import datetime, timezone
import os
from pathlib import Path
import sqlite3
import threading


MAX_HISTORY_ENTRIES = 500


def default_history_path() -> Path:
    data_home = Path(os.environ.get("XDG_DATA_HOME", Path.home() / ".local" / "share"))
    return data_home / "flow-linux" / "history.db"


@dataclass(frozen=True)
class HistoryEntry:
    id: int
    created_at: str
    transcript: str
    duration: float
    provider: str
    model: str
    insertion_status: str


class HistoryStore:
    """Open short-lived SQLite connections so worker and GTK threads can share storage."""

    def __init__(self, path=None, max_entries=MAX_HISTORY_ENTRIES):
        if isinstance(max_entries, bool) or int(max_entries) < 1:
            raise ValueError("max_entries must be a positive integer")
        self.path = Path(path) if path is not None else default_history_path()
        self.max_entries = int(max_entries)
        self._lock = threading.RLock()
        self._initialized = False

    def _connect(self):
        self.path.parent.mkdir(mode=0o700, parents=True, exist_ok=True)
        try:
            self.path.parent.chmod(0o700)
        except OSError:
            pass
        connection = sqlite3.connect(self.path, timeout=5.0)
        connection.row_factory = sqlite3.Row
        connection.execute("PRAGMA busy_timeout = 5000")
        try:
            self.path.chmod(0o600)
        except OSError:
            pass
        return connection

    def _initialize_database(self):
        connection = self._connect()
        try:
            integrity = connection.execute("PRAGMA quick_check").fetchone()[0]
            if integrity != "ok":
                raise sqlite3.DatabaseError(f"database integrity check failed: {integrity}")
            connection.executescript(
                """
                CREATE TABLE IF NOT EXISTS dictations (
                    id INTEGER PRIMARY KEY AUTOINCREMENT,
                    created_at TEXT NOT NULL,
                    transcript TEXT NOT NULL,
                    duration REAL NOT NULL CHECK (duration >= 0),
                    provider TEXT NOT NULL,
                    model TEXT NOT NULL,
                    insertion_status TEXT NOT NULL
                        CHECK (insertion_status IN ('pending', 'inserted', 'failed'))
                );
                CREATE INDEX IF NOT EXISTS dictations_created_at
                    ON dictations(created_at DESC, id DESC);
                """
            )
            connection.commit()
        finally:
            connection.close()
        try:
            self.path.chmod(0o600)
        except OSError:
            pass

    @staticmethod
    def _is_corrupt_error(error):
        message = str(error).lower()
        return any(phrase in message for phrase in (
            "malformed", "not a database", "database disk image", "file is encrypted",
        ))

    def _preserve_corrupt_database(self):
        stamp = datetime.now(timezone.utc).strftime("%Y%m%dT%H%M%S%fZ")
        damaged = self.path.with_name(f"{self.path.name}.corrupt-{stamp}")
        os.replace(self.path, damaged)
        # Preserve any SQLite sidecars alongside the damaged file for possible
        # manual recovery; normal operation never creates audio files here.
        for suffix in ("-wal", "-shm"):
            sidecar = Path(f"{self.path}{suffix}")
            if sidecar.exists():
                os.replace(sidecar, Path(f"{damaged}{suffix}"))

    def _ensure_initialized(self):
        with self._lock:
            if self._initialized:
                return
            try:
                self._initialize_database()
            except sqlite3.DatabaseError as exc:
                if not self.path.exists() or not self._is_corrupt_error(exc):
                    raise
                self._preserve_corrupt_database()
                self._initialize_database()
            self._initialized = True

    @staticmethod
    def _entry(row):
        return HistoryEntry(
            id=row["id"], created_at=row["created_at"],
            transcript=row["transcript"], duration=row["duration"],
            provider=row["provider"], model=row["model"],
            insertion_status=row["insertion_status"],
        )

    def add(self, transcript, duration, provider, model):
        if not isinstance(transcript, str) or not transcript.strip():
            raise ValueError("transcript must contain text")
        if float(duration) < 0:
            raise ValueError("duration cannot be negative")
        self._ensure_initialized()
        created_at = datetime.now(timezone.utc).isoformat(timespec="microseconds")
        with self._lock:
            connection = self._connect()
            try:
                connection.execute("BEGIN IMMEDIATE")
                cursor = connection.execute(
                    """INSERT INTO dictations
                       (created_at, transcript, duration, provider, model, insertion_status)
                       VALUES (?, ?, ?, ?, ?, 'pending')""",
                    (created_at, transcript, float(duration), str(provider), str(model)),
                )
                entry_id = cursor.lastrowid
                connection.execute(
                    """DELETE FROM dictations WHERE id NOT IN (
                           SELECT id FROM dictations
                           ORDER BY created_at DESC, id DESC LIMIT ?
                       )""",
                    (self.max_entries,),
                )
                connection.commit()
                return entry_id
            except Exception:
                connection.rollback()
                raise
            finally:
                connection.close()

    def mark_insertion(self, entry_id, status):
        if status not in {"pending", "inserted", "failed"}:
            raise ValueError("insertion status must be pending, inserted, or failed")
        self._ensure_initialized()
        with self._lock:
            connection = self._connect()
            try:
                with connection:
                    connection.execute(
                        "UPDATE dictations SET insertion_status = ? WHERE id = ?",
                        (status, int(entry_id)),
                    )
            finally:
                connection.close()

    def list_entries(self, search="", limit=MAX_HISTORY_ENTRIES):
        self._ensure_initialized()
        try:
            limit = max(1, min(int(limit), MAX_HISTORY_ENTRIES))
        except (TypeError, ValueError):
            limit = MAX_HISTORY_ENTRIES
        search = str(search or "").strip()
        pattern = ("%" + search.replace("\\", "\\\\").replace("%", "\\%").replace("_", "\\_") + "%")
        with self._lock:
            connection = self._connect()
            try:
                rows = connection.execute(
                    """SELECT id, created_at, transcript, duration, provider, model,
                              insertion_status
                       FROM dictations
                       WHERE ? = '' OR transcript LIKE ? ESCAPE '\\'
                       ORDER BY created_at DESC, id DESC LIMIT ?""",
                    (search, pattern, limit),
                ).fetchall()
                return [self._entry(row) for row in rows]
            finally:
                connection.close()

    def delete(self, entry_id):
        self._ensure_initialized()
        with self._lock:
            connection = self._connect()
            try:
                with connection:
                    cursor = connection.execute("DELETE FROM dictations WHERE id = ?", (int(entry_id),))
                    return cursor.rowcount > 0
            finally:
                connection.close()

    def clear(self):
        self._ensure_initialized()
        with self._lock:
            connection = self._connect()
            try:
                with connection:
                    cursor = connection.execute("DELETE FROM dictations")
                    return cursor.rowcount
            finally:
                connection.close()

File: src-tauri/src/history.rs
use rusqlite::{params, Connection};
use serde::{Deserialize, Serialize};
#[cfg(test)]
use std::fs;
use std::{
    path::{Path, PathBuf},
    time::Duration,
};

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct Entry {
    pub id: i64,
    pub created_at: String,
    pub transcript: String,
    pub duration: f64,
    pub provider: String,
    pub model: String,
    pub insertion_status: String,
}
fn database() -> PathBuf {
    crate::config::data_dir().join("history.db")
}
fn connect(path: &Path) -> Result<Connection, String> {
    crate::storage::secure_directory(path.parent().ok_or("Invalid history path")?)?;
    let connection = Connection::open(path).map_err(|e| e.to_string())?;
    crate::storage::secure_file(path)?;
    connection
        .busy_timeout(Duration::from_secs(5))
        .map_err(|e| e.to_string())?;
    connection.execute_batch("CREATE TABLE IF NOT EXISTS dictations (id INTEGER PRIMARY KEY AUTOINCREMENT, created_at TEXT NOT NULL, transcript TEXT NOT NULL, duration REAL NOT NULL CHECK(duration >= 0), provider TEXT NOT NULL, model TEXT NOT NULL, insertion_status TEXT NOT NULL CHECK(insertion_status IN ('pending','inserted','failed'))); CREATE INDEX IF NOT EXISTS dictations_created_at ON dictations(created_at DESC,id DESC);").map_err(|e| e.to_string())?;
    Ok(connection)
}
pub fn add(text: &str, duration: f64, model: &str) -> Result<i64, String> {
    add_at(&database(), text, duration, model)
}
fn add_at(path: &Path, text: &str, duration: f64, model: &str) -> Result<i64, String> {
    let mut connection = connect(path)?;
    let transaction = connection.transaction().map_err(|e| e.to_string())?;
    transaction.execute("INSERT INTO dictations(created_at,transcript,duration,provider,model,insertion_status) VALUES(strftime('%Y-%m-%dT%H:%M:%fZ','now'),?1,?2,'Groq',?3,'pending')", params![text,duration.max(0.0),model]).map_err(|e| e.to_string())?;
    let id = transaction.last_insert_rowid();
    transaction.execute("DELETE FROM dictations WHERE id NOT IN (SELECT id FROM dictations ORDER BY created_at DESC,id DESC LIMIT 500)", []).map_err(|e| e.to_string())?;
    transaction.commit().map_err(|e| e.to_string())?;
    Ok(id)
}
pub fn mark(id: i64, status: &str) {
    let _ = mark_at(&database(), id, status);
}
fn mark_at(path: &Path, id: i64, status: &str) -> Result<(), String> {
    connect(path)?
        .execute(
            "UPDATE dictations SET insertion_status=?1 WHERE id=?2",
            params![status, id],
        )
        .map_err(|e| e.to_string())?;
    Ok(())
}
pub fn list(search: &str) -> Result<Vec<Entry>, String> {
    list_at(&database(), search)
}
fn list_at(path: &Path, search: &str) -> Result<Vec<Entry>, String> {
    let connection = connect(path)?;
    let mut statement = connection.prepare("SELECT id,created_at,transcript,duration,provider,model,insertion_status FROM dictations WHERE instr(lower(transcript),lower(?1))>0 ORDER BY created_at DESC,id DESC LIMIT 500").map_err(|e| e.to_string())?;
    let rows = statement
        .query_map([search], |row| {
            Ok(Entry {
                id: row.get(0)?,
                created_at: row.get(1)?,
                transcript: row.get(2)?,
                duration: row.get(3)?,
                provider: row.get(4)?,
                model: row.get(5)?,
                insertion_status: row.get(6)?,
            })
        })
        .map_err(|e| e.to_string())?;
    rows.collect::<Result<Vec<_>, _>>()
        .map_err(|e| e.to_string())
}
pub fn delete(id: i64) -> Result<(), String> {
    delete_at(&database(), id)
}
fn delete_at(path: &Path, id: i64) -> Result<(), String> {
    connect(path)?
        .execute("DELETE FROM dictations WHERE id=?1", [id])
        .map_err(|e| e.to_string())?;
    Ok(())
}
pub fn clear() -> Result<(), String> {
    clear_at(&database())
}
fn clear_at(path: &Path) -> Result<(), String> {
    connect(path)?
        .execute("DELETE FROM dictations", [])
        .map_err(|e| e.to_string())?;
    Ok(())
}

#[cfg(test)]
mod tests {
    use super::*;
    #[test]
    fn local_history_supports_insert_search_status_and_delete() {
        let folder = std::env::temp_dir().join(format!(
            "flow-history-test-{}-{}",
            std::process::id(),
            std::time::SystemTime::now()
                .duration_since(std::time::UNIX_EPOCH)
                .unwrap()
                .as_nanos()
        ));
        fs::create_dir(&folder).unwrap();
        crate::storage::secure_directory(&folder).unwrap();
        let path = folder.join("history.db");
        let text = "Flow's dictation; DROP TABLE dictations; --";
        let id = add_at(&path, text, 1.25, "whisper-test").unwrap();
        let found = list_at(&path, "DROP TABLE").unwrap();
        assert_eq!(found.len(), 1);
        assert_eq!(found[0].transcript, text);
        mark_at(&path, id, "inserted").unwrap();
        assert_eq!(list_at(&path, "").unwrap()[0].insertion_status, "inserted");
        delete_at(&path, id).unwrap();
        assert!(list_at(&path, "").unwrap().is_empty());
        clear_at(&path).unwrap();
        fs::remove_dir_all(folder).unwrap();
    }
}
